tools/src/main/java/org/apache/kafka/tools/ConsumerPerformance.java

```diff
@@ -0,0 +1,399 @@
+            System.err.println(e.getMessage());
+            System.err.println(Utils.stackTrace(e));
+            Exit.exit(1);
+        }
+    }
+
+    protected static void printHeader(boolean showDetailedStats) {
+        String newFieldsInHeader = ", rebalance.time.ms, fetch.time.ms, fetch.MB.sec, fetch.nMsg.sec";
+        if (!showDetailedStats)
+            System.out.printf("start.time, end.time, data.consumed.in.MB, MB.sec, data.consumed.in.nMsg, nMsg.sec%s%n", newFieldsInHeader);
+        else
+            System.out.printf("time, threadId, data.consumed.in.MB, MB.sec, data.consumed.in.nMsg, nMsg.sec%s%n", newFieldsInHeader);
+    }
+
+    @SuppressWarnings("unchecked")
+    private static void consume(KafkaConsumer<byte[], byte[]> consumer,
+                                ConsumerPerfOptions options,
+                                AtomicLong totalMessagesRead,
+                                AtomicLong totalBytesRead,
+                                AtomicLong joinGroupTimeMs,
+                                long bytesRead,
+                                long messagesRead,
+                                long lastBytesRead,
+                                long lastMessagesRead,
+                                long joinStartMs,
+                                long joinTimeMsInSingleRound) {
+        long numMessages = options.numMessages();
+        long recordFetchTimeoutMs = options.recordFetchTimeoutMs();
+        long reportingIntervalMs = options.reportingIntervalMs();
+        boolean showDetailedStats = options.showDetailedStats();
+        SimpleDateFormat dateFormat = options.dateFormat();
+        consumer.subscribe(options.topic(),
+            new ConsumerPerfRebListener(joinGroupTimeMs, joinStartMs, joinTimeMsInSingleRound));
+
+        // now start the benchmark
+        long currentTimeMs = System.currentTimeMillis();
+        long lastReportTimeMs = currentTimeMs;
+        long lastConsumedTimeMs = currentTimeMs;
+
+        while (messagesRead < numMessages && currentTimeMs - lastConsumedTimeMs <= recordFetchTimeoutMs) {
+            ConsumerRecords<byte[], byte[]> records = consumer.poll(Duration.ofMillis(100));
+            currentTimeMs = System.currentTimeMillis();
+            if (!records.isEmpty())
+                lastConsumedTimeMs = currentTimeMs;
+            for (ConsumerRecord<byte[], byte[]> record : records) {
+                messagesRead += 1;
+                if (record.key() != null)
+                    bytesRead += record.key().length;
+                if (record.value() != null)
+                    bytesRead += record.value().length;
+                if (currentTimeMs - lastReportTimeMs >= reportingIntervalMs) {
+                    if (showDetailedStats)
+                        printConsumerProgress(0, bytesRead, lastBytesRead, messagesRead, lastMessagesRead,
+                            lastReportTimeMs, currentTimeMs, dateFormat, joinTimeMsInSingleRound);
```

Comment:
This `joinTimeMsInSingleRound` will not be updated by the custom rebalance listener, is it still correct?

Reply:
This is only related to the extended class, but you are right, we need to use AtomicLong passed by reference here. There is also another issue which I'm going to fix.